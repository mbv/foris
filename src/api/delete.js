/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useReducer, useCallback } from "react";
import axios from "axios";

import {
    API_ACTIONS, TIMEOUT, HEADERS, APIReducer, getErrorMessage,
} from "./utils";

export function useAPIDelete(url) {
    const [state, dispatch] = useReducer(APIReducer, {
        isSending: false,
        isError: false,
        isSuccess: false,
        data: null,
    });

    const requestDelete = useCallback(async () => {
        dispatch({ type: API_ACTIONS.INIT });
        try {
            await axios.delete(url, {
                timeout: TIMEOUT,
                headers: HEADERS,
            });
            dispatch({ type: API_ACTIONS.SUCCESS });
        } catch (error) {
            dispatch({
                type: API_ACTIONS.FAILURE,
                payload: getErrorMessage(error),
                status: error.response.status,
            });
        }
    }, [url]);

    return [state, requestDelete];
}
