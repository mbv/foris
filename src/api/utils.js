/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import axios from "axios";

export const HEADERS = {
    Accept: "application/json",
    "Content-Type": "application/json",
    "X-CSRFToken": getCookie("_csrf_token"),
};

export const TIMEOUT = 30500;

export const API_ACTIONS = {
    INIT: 1,
    SUCCESS: 2,
    FAILURE: 3,
};

export const API_STATE = {
    INIT: "init",
    SENDING: "sending",
    SUCCESS: "success",
    ERROR: "error",
};

export const API_METHODS = {
    GET: axios.get,
    POST: axios.post,
    PATCH: axios.patch,
    PUT: axios.put,
    DELETE: axios.delete,
};

function getCookie(name) {
    let cookieValue = null;
    if (document.cookie && document.cookie !== "") {
        const cookies = document.cookie.split(";");
        for (let i = 0; i < cookies.length; i++) {
            const cookie = cookies[i].trim();
            // Does this cookie string begin with the name we want?
            if (cookie.substring(0, name.length + 1) === (`${name}=`)) {
                cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
                break;
            }
        }
    }
    return cookieValue;
}

export function getErrorPayload(error) {
    if (error.response) {
        if (error.response.status === 403) {
            return _("The session is expired. Please log in again.");
        }
        return getJSONErrorMessage(error);
    }
    if (error.code === "ECONNABORTED") {
        return _("Timeout error occurred.");
    }
    if (error.request) {
        return _("No response received.");
    }
    /* eslint no-console: "off" */
    console.error(error);
    return _("An unknown error occurred. Check the console for more info.");
}

export function getJSONErrorMessage(error) {
    if (error.response.headers["content-type"] === "application/json") {
        return error.response.data;
    }
    return _("An unknown API error occurred.");
}
