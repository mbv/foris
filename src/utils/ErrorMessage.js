/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

ErrorMessage.propTypes = {
    message: PropTypes.string,
};

ErrorMessage.defaultProps = {
    message: _("An error occurred while fetching data."),
};

export function ErrorMessage({ message }) {
    return (
        <p className="text-center text-danger">{message}</p>
    );
}
