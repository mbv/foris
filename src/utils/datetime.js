import moment from "moment";

export function toLocaleDateString(date, { inputFormat, outputFormat = "LLL" } = {}) {
    const parsedDate = inputFormat ? moment(date, inputFormat) : moment(date);
    return parsedDate
        .locale(ForisTranslations.locale)
        .format(outputFormat);
}
