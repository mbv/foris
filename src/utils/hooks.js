/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useState, useEffect } from "react";

/** Execute callback when condition is set to true. */
export function useConditionalTimeout({ callback, timeout = 125 }, ...callbackArgs) {
    const [condition, setCondition] = useState(false);
    useEffect(() => {
        if (condition) {
            const interval = setTimeout(() => callback(...callbackArgs), timeout);
            return () => setTimeout(interval);
        }
    }, [condition, callback, timeout, callbackArgs]);
    return setCondition;
}
