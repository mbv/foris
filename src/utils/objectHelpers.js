/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/** Return undefined if object has no keys, otherwise return object. */
export function undefinedIfEmpty(instance) {
    if (Object.keys(instance).length > 0) {
        return instance;
    }
    return undefined;
}

/** Return object without keys that have undefined value. */
export function withoutUndefinedKeys(instance) {
    return Object.keys(instance).reduce(
        (accumulator, key) => {
            if (instance[key] !== undefined) {
                accumulator[key] = instance[key];
            }
            return accumulator;
        },
        {},
    );
}
