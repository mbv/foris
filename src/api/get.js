/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useReducer, useCallback } from "react";
import axios from "axios";

import { ForisURLs } from "forisUrls";
import { API_ACTIONS, TIMEOUT } from "./utils";

const APIGetReducer = (state, action) => {
    switch (action.type) {
    case API_ACTIONS.INIT:
        return {
            ...state,
            isLoading: true,
            isError: false,
        };
    case API_ACTIONS.SUCCESS:
        return {
            ...state,
            isLoading: false,
            isError: false,
            data: action.payload,
        };
    case API_ACTIONS.FAILURE:
        if (action.status === 403) window.location.assign(ForisURLs.login);
        return {
            ...state,
            isLoading: false,
            isError: true,
            data: action.payload,
        };
    default:
        throw new Error();
    }
};

export function useAPIGet(url) {
    const [state, dispatch] = useReducer(APIGetReducer, {
        isLoading: false,
        isError: false,
        data: null,
    });
    const get = useCallback(async () => {
        dispatch({ type: API_ACTIONS.INIT });
        try {
            const result = await axios.get(url, {
                timeout: TIMEOUT,
            });
            dispatch({ type: API_ACTIONS.SUCCESS, payload: result.data });
        } catch (error) {
            dispatch({
                type: API_ACTIONS.FAILURE,
                payload: error.response.data,
                status: error.response.status,
            });
        }
    }, [url]);

    return [state, get];
}
