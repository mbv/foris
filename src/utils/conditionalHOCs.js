/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { Spinner } from "bootstrap/Spinner";
import { API_STATE } from "api/utils";
import { ErrorMessage } from "./ErrorMessage";

function withEither(conditionalFn, Either) {
    return (Component) => (props) => {
        if (conditionalFn(props)) {
            return <Either />;
        }
        return <Component {...props} />;
    };
}

// Loading

function isSending(props) {
    if (Array.isArray(props.apiState)) {
        return props.apiState.some(
            (state) => [API_STATE.INIT, API_STATE.SENDING].includes(state),
        );
    }
    return [API_STATE.INIT, API_STATE.SENDING].includes(props.apiState);
}

const withSpinner = (conditionalFn) => withEither(conditionalFn, Spinner);
const withSending = (Either) => withEither(isSending, Either);
const withSpinnerOnSending = withSpinner(isSending);

// Error handling

const withError = (conditionalFn) => withEither(conditionalFn, ErrorMessage);
const withErrorMessage = withError(
    (props) => {
        if (Array.isArray(props.apiState)) {
            return props.apiState.includes(API_STATE.ERROR);
        }
        return props.apiState === API_STATE.ERROR;
    },
);

export {
    withEither, withSpinner, withSending, withSpinnerOnSending, withError, withErrorMessage,
};
