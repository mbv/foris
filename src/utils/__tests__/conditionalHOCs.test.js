/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render } from "customTestRender";
import { API_STATE } from "api/utils";
import {
    withEither, withSpinner, withSending, withSpinnerOnSending, withError, withErrorMessage,
} from "../conditionalHOCs";

describe("conditional HOCs", () => {
    const First = () => <p>First</p>;
    const Alternative = () => <p>Alternative</p>;

    describe("withEither", () => {
        it("should render First component", () => {
            const withAlternative = withEither(() => false, Alternative);
            const FirstWithConditional = withAlternative(First);
            const { getByText } = render(<FirstWithConditional />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render Alternative component", () => {
            const withAlternative = withEither(() => true, Alternative);
            const FirstWithConditional = withAlternative(First);
            const { getByText } = render(<FirstWithConditional />);
            expect(getByText("Alternative")).toBeDefined();
        });
    });

    describe("withSpinner", () => {
        it("should render First component", () => {
            const withSpinnerHidden = withSpinner(() => false);
            const FirstWithConditional = withSpinnerHidden(First);
            const { getByText } = render(<FirstWithConditional />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render spinner", () => {
            const withSpinnerVisible = withSpinner(() => true);
            const FirstWithConditional = withSpinnerVisible(First);
            const { container } = render(<FirstWithConditional />);
            expect(container).toMatchSnapshot();
        });
    });

    describe("withSending", () => {
        it("should render First component", () => {
            const withAlternative = withSending(Alternative);
            const FirstWithConditional = withAlternative(First);
            const { getByText } = render(<FirstWithConditional apiState={API_STATE.SUCCESS} />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render Alternative component", () => {
            const withAlternative = withSending(Alternative);
            const FirstWithConditional = withAlternative(First);
            const { getByText } = render(<FirstWithConditional apiState={API_STATE.SENDING} />);
            expect(getByText("Alternative")).toBeDefined();
        });
    });

    describe("withSpinnerOnSending", () => {
        it("should render First component", () => {
            const FirstWithConditional = withSpinnerOnSending(First);
            const { getByText } = render(<FirstWithConditional apiState={API_STATE.SUCCESS} />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render spinner", () => {
            const FirstWithConditional = withSpinnerOnSending(First);
            const { container } = render(<FirstWithConditional apiState={API_STATE.SENDING} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe("withError", () => {
        it("should render First component", () => {
            const withErrorHidden = withError(() => false);
            const FirstWithConditional = withErrorHidden(First);
            const { getByText } = render(<FirstWithConditional />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render error message", () => {
            const withErrorVisible = withError(() => true);
            const FirstWithConditional = withErrorVisible(First);
            const { container } = render(<FirstWithConditional />);
            expect(container).toMatchSnapshot();
        });
    });

    describe("withErrorMessage", () => {
        it("should render First component", () => {
            const FirstWithConditional = withErrorMessage(First);
            const { getByText } = render(<FirstWithConditional apiState={API_STATE.SUCCESS} />);
            expect(getByText("First")).toBeDefined();
        });

        it("should render error message", () => {
            const FirstWithConditional = withErrorMessage(First);
            const { container } = render(<FirstWithConditional apiState={API_STATE.ERROR} />);
            expect(container).toMatchSnapshot();
        });
    });
});
