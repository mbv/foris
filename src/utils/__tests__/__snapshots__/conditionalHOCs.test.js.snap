// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditional HOCs withError should render error message 1`] = `
<div>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`conditional HOCs withErrorMessage should render error message 1`] = `
<div>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`conditional HOCs withSpinner should render spinner 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`conditional HOCs withSpinnerOnSending should render spinner 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
